bionic/deriver.py

```diff
@@ -460,46 +411,227 @@ def is_resolved(self):
+        assert in_progress_entry.stage == EntryStage.ACTIVE
+        assert in_progress_entry.future is None
+        assert in_progress_entry not in self._in_progress_entries
+
+        in_progress_entry.stage = EntryStage.IN_PROGRESS
+        in_progress_entry.future = future
+        self._in_progress_entries[
+            in_progress_entry.state.task.keys[0]
+        ] = in_progress_entry
+
+    def _mark_entry_completed(self, completed_entry):
+        assert completed_entry.stage == EntryStage.ACTIVE or EntryStage.IN_PROGRESS
 
-        for completed_tk in completed_state.task.keys:
+        if completed_entry.stage == EntryStage.IN_PROGRESS:
+            completed_entry.future = None
+            del self._in_progress_entries[completed_entry.state.task.keys[0]]
+
+        completed_entry.stage = EntryStage.COMPLETED
+        self._unblock_entries(completed_entry)
+
+    def _ready_blocking_entries(self, blocked_entry):
+        incomplete_dep_states = blocked_entry.state.incomplete_dep_states()
+        for incomplete_dep_state in incomplete_dep_states:
+            incomplete_entry = self._get_or_create_entry_for_state(incomplete_dep_state)
+            if incomplete_entry.stage == EntryStage.NEW:
+                self._mark_entry_ready(incomplete_entry)
+
+    def _unblock_entries(self, completed_entry):
+        for completed_tk in completed_entry.state.task.keys:
             affected_blockages = self._blockage_lists_by_blocking_tk[completed_tk]
             for blockage in affected_blockages:
                 blockage.mark_task_key_complete(completed_tk)
-                if blockage.is_resolved():
-                    yield blockage.blocked_state
+                if (
+                    blockage.is_resolved()
+                    and blockage.blocked_entry.stage == EntryStage.BLOCKED
+                ):
+                    self._mark_entry_ready(blockage.blocked_entry)
 
-    def get_all_blocked_states(self):
+    def _get_all_blocked_entries(self):
         return {
-            blockage.blocked_state
+            blockage.blocked_entry
             for blockages in self._blockage_lists_by_blocking_tk.values()
             for blockage in blockages
             if not blockage.is_resolved()
         }
+
+
+class TaskRunnerEntry:
+    """
+    Basic unit of `TaskCompletionRunner` that contains the data for
+    `TaskState` execution and tracking.
+    """
+
+    def __init__(self, state):
+        self.state = state
+        self.stage = EntryStage.NEW
+        self.future = None
+
+
+class EntryStage(Enum):
+    """
+    Represents the stage of a `TaskRunnerEntry`.
+    """
+
+    """
+    Entry is created and hasn't been processed.
+    This is the first stage in processing.
+    Valid next stages: [READY]
+    """
+    NEW = auto()
+
+    """
+    Entry is ready to be processed.
+    Valid next stages: [ACTIVE]
+    """
+    READY = auto()
+
+    """
+    Entry is actively running.
```

Comment:
"running" doesn't sound quite right to me, since the entry could end up being blocked. Could we say something like "We are actively attempting to start running this entry."?